File: graphics/curves/4/eric_kolbusz/parser.py
from display import *
from matrix import *
from draw import *

def parse_file( fname, points, transform ):
    screen = new_screen()
    color = [255,255,255]
    f = open(fname, 'r')
    lines = f.readlines()
    for i in range(len(lines)):
        line = lines[i]
        line = line.strip()
        if line == 'a':
            points = matrix_mult(transform, points)
        elif line == 'i':
            ident(transform)
        elif line == 'v':
            screen = new_screen()
            draw_lines(points, screen, color)
            display(screen)
        elif line in list('lstxyzcbhg'):
            args = lines[i+1].strip().split(' ')
            if line == 'l':
                try:
                    args = [int(a) for a in args]
                    add_edge(points, args[0], args[1], args[2], args[3], args[4], args[5])
                except:
                    print "Invalid args for operation l, exiting"
                    exit(1)
            if line == 's':
                try:
                    args = [float(a) for a in args]
                    m = make_scale(args[0], args[1], args[2])
                except:
                    print "Invald args for operation s, exiting"
                    exit(1)
                transform = matrix_mult(m, transform)
            if line == 't':
                try:
                    args = [int(a) for a in args]
                    m = make_translate(args[0], args[1], args[2])
                except:
                    print "Invald args for operation t, exiting"
                    exit(1)   
                transform = matrix_mult(m, transform)
            if line == 'x':
                try:
                    args = [int(a) for a in args]
                    m = make_rotX(args[0])
                except:
                    print "Invald args for operation x, exiting"
                    exit(1)   
                transform = matrix_mult(m, transform)
            if line == 'y':
                try:
                    args = [int(a) for a in args]
                    m = make_rotY(args[0])
                except:
                    print "Invald args for operation y, exiting"
                    exit(1)   
                transform = matrix_mult(m, transform)
            if line == 'z':
                try:
                    args = [int(a) for a in args]
                    m = make_rotZ(args[0])
                except:
                    print "Invald args for operation z, exiting"
                    exit(1)   
                transform = matrix_mult(m, transform)
            if line == 'c':
                try:
                    args = [int(a) for a in args]
                    add_circle(points, args[0], args[1], 0, args[2], .001)
                except:
                    print "Invald args for operation c, exiting"
                    exit(1)
            if line in list('bh'):
                try:
                    args = [int(a) for a in args]
                    add_curve(points, args[0], args[1], args[2], args[3], args[4], args[5], args[6], args[7], .001, line)
                except:
                    print "Invald args for operation %s, exiting"%(line)
                    exit(1)
            if line == 'g':
                draw_lines(points, screen, color)
                #save_ppm(screen, args[0])
                save_extension(screen, args[0])
            
                 


points = []
transform = new_matrix()

parse_file( 'script_c', points, transform )

File: graphics/curves/4/lise_ho/parser.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "ml6.h"
#include "display.h"
#include "draw.h"
#include "matrix.h"
#include "parser.h"


/*======== void parse_file () ==========
Inputs:   char * filename 
          struct matrix * transform, 
          struct matrix * pm,
          screen s
Returns: 

Goes through the file named filename and performs all of the actions listed in that file.
The file follows the following format:
     Every command is a single character that takes up a line
     Any command that requires arguments must have those arguments in the second line.
     The commands are as follows:
         l: add a line to the edge matrix - 
	    takes 6 arguemnts (x0, y0, z0, x1, y1, z1)
	 c: add a circle to the edge matrix - 
	    takes 3 arguments (cx, cy, r)
	 h: add a hermite curve to the edge matrix -
	    takes 8 arguments (x0, y0, x1, y1, x2, y2, x3, y3)
	 b: add a bezier curve to the edge matrix -
	    takes 8 arguments (x0, y0, x1, y1, x2, y2, x3, y3)
	 i: set the transform matrix to the identity matrix - 
	 s: create a scale matrix, 
	    then multiply the transform matrix by the scale matrix - 
	    takes 3 arguments (sx, sy, sz)
	 t: create a translation matrix, 
	    then multiply the transform matrix by the translation matrix - 
	    takes 3 arguments (tx, ty, tz)
	 x: create an x-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 y: create an y-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 z: create an z-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 a: apply the current transformation matrix to the 
	    edge matrix
	 v: draw the lines of the edge matrix to the screen
	    display the screen
	 g: draw the lines of the edge matrix to the screen
	    save the screen to a file -
	    takes 1 argument (file name)
	 q: end parsing

See the file script for an example of the file format


IMPORTANT MATH NOTE:
the trig functions int math.h use radian mesure, but us normal
humans use degrees, so the file will contain degrees for rotations,
be sure to conver those degrees to radians (M_PI is the constant
for PI)
====================*/
void parse_file ( char * filename, 
                  struct matrix * transform, 
                  struct matrix * pm,
                  screen s) {

  FILE *f;
  char line[256];
  struct matrix * tmp;
  double angle;
  color g;

  g.red = 255;
  g.green = 0;
  g.blue = 255;
  
  clear_screen(s);

  if ( strcmp(filename, "stdin") == 0 ) 
    f = stdin;
  else
    f = fopen(filename, "r");
  
  while ( fgets(line, 255, f) != NULL ) {
    line[strlen(line)-1]='\0';
    //printf(":%s:\n",line);
    char c;
    double x, y, z, x1, y1, z1, x2, y2, z2, x3, y3, z3, x4, y4, z4;
   
    c = line[0];

    switch (c) {
    case 'l':
      //      printf("LINE!\n");
      fgets(line, 255, f);
      //      printf("\t%s", line);
      //line[strlen(line)-1]='\0';
      sscanf(line, "%lf %lf %lf %lf %lf %lf", &x, &y, &z, &x1, &y1, &z1);
      add_edge(pm, x, y, z, x1, y1, z1);
      // printf( "%lf %lf %lf %lf %lf %lf\n", x, y, z, x1, y1, z1);
      break;
    case 's':
      //printf("SCALE\n");
      fgets(line, 255, f);
      //line[strlen(line)-1]='\0';      
      sscanf(line, "%lf %lf %lf", &x, &y, &z);
      tmp = make_scale(x, y, z);
      matrix_mult(tmp, transform);
      //print_matrix(transform);
      break;
    case 't':
      //printf("TRANSLATE\n");
      fgets(line, 255, f);
      //      line[strlen(line)-1]='\0';      
      sscanf(line, "%lf %lf %lf", &x, &y, &z);
      tmp = make_translate(x, y, z);
      matrix_mult(tmp, transform);
      //print_matrix(transform);
      break;
    case 'x':
      //printf("ROTATE!\n");
      fgets(line, 255, f);
      sscanf(line, "%lf", &angle);
      angle = angle * (M_PI/180);
      tmp = make_rotX( angle);
      matrix_mult(tmp, transform);
      break;
    case 'y':
      //printf("ROTATE!\n");
      fgets(line, 255, f);
      sscanf(line, "%lf", &angle);
      angle = angle * (M_PI / 180);
      tmp = make_rotY( angle);
      matrix_mult(tmp, transform);
      break;
    case 'z':
      //printf("ROTATE!\n");
      fgets(line, 255, f);
      sscanf(line, "%lf", &angle);
      angle = angle * (M_PI / 180);
      tmp = make_rotZ( angle);
      matrix_mult(tmp, transform);
      break;
    case 'i':
      ident(transform);
      break;
    case 'a':
      //printf("APPLY!\n");
      //print_matrix( transform );
      //      print_matrix(pm);
      matrix_mult(transform, pm);
      break;
    case 'v':
      clear_screen(s);
      draw_lines(pm, s, g);
      display(s);
      break;
    case 'g':
      fgets(line, 255, f);
      // line[strlen(line)-1] = '\0';
      clear_screen(s);
      draw_lines(pm, s, g);
      save_extension(s, line);
      break;
    case 'q':
      return;
    case 'c':
      //circle
      /* c: add a circle to the edge matrix - 
	    takes 3 arguments (cx, cy, r)
      */
      fgets(line, 255, f);
      //      line[strlen(line)-1]='\0';      
      double cx, cy, r;
      sscanf(line, "%lf %lf %lf", &cx, &cy, &r);

      printf("%f   %f  %f \n", cx,cy,r);
      // add circle routine  - how many lines are we planning to draw?
      add_circle(pm, cx, cy, r, 50);
      //printf("ADSLKJADSFLKJADSLFKJDSA\n\n\n");
      break;
    case 'h':
      //hermit curve
      /*  h: add a hermite curve to the edge matrix -
	    takes 8 arguments (x0, y0, x1, y1, x2, y2, x3, y3)
	    //remember the curve is between points (x0, y0) and (x2, y2).
	    //(x1, y1) and (x3, y3) are used to calculate the rates of change at the endpoints
      */
      fgets(line, 255, f);
      //      line[strlen(line)-1]='\0';      
      sscanf(line, "%lf %lf %lf %lf %lf %lf %lf %lf", &x, &y, &x1, &y1, &x2, &y2, &x3, &y3);
      // add hermit curve routine
      add_curve(pm,x,y,x1,y1,x2,y2,x3,y3,0.01,1);
      //   return;
      printf("adsfkljads");
      break;
    case 'b':
      //bezier curve
      /*    b: add a bezier curve to the edge matrix -
	    takes 8 arguments (x0, y0, x1, y1, x2, y2, x3, y3)
	    //This curve is drawn between (x0, y0) and (x3, y3)
      */
      fgets(line, 255, f);
      //      line[strlen(line)-1]='\0';      
      sscanf(line, "%lf %lf %lf %lf %lf %lf %lf %lf", &x, &y, &x1, &y1, &x2, &y2, &x3, &y3);
      //add bezier curve stuf
      add_curve(pm,x,y,x1,y1,x2,y2,x3,y3,0.02,2);
      break;
   
    default:
      printf("Invalid command\n");
      break;
    }
  }

  free_matrix(tmp);
  fclose(f);
  //printf("END PARSE\n");
}

  
